storage/transfermanager/downloader.go

```diff
@@ -0,0 +1,149 @@
+package transfermanager
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"cloud.google.com/go/storage"
+)
+
+type TransferManagerOption interface {
+	apply(*transferManagerConfig)
+}
+
+func WithWorkers(numWorkers int) TransferManagerOption {
+	return &withWorkers{numWorkers: numWorkers}
+}
+
+type withWorkers struct {
+	numWorkers int
+}
+
+func (ww withWorkers) apply(tm *transferManagerConfig) {
+	tm.numWorkers = ww.numWorkers
+}
+
+func WithPartSize(partSize int) TransferManagerOption {
+	return &withPartSize{partSize: partSize}
+}
+
+type withPartSize struct {
+	partSize int
+}
+
+func (wps withPartSize) apply(tm *transferManagerConfig) {
+	tm.partSize = wps.partSize
+}
+
+// etc for additional options.
+
+type transferManagerConfig struct {
+	// Workers in thread pool; default numCPU/2 based on previous benchmarks?
+	numWorkers int
+	// Size of shards to transfer; Python found 32 MiB to be good default for
+	// JSON downloads but gRPC may benefit from larger.
+	partSize int
+	// Timeout for a single operation (including all retries).
+	perOperationTimeout time.Duration
+
+	// others, including opts that apply to uploads.
+}
+
+// etc
+
+type Downloader struct {
+	ctx             context.Context
```

Comment:
Do we want this overarching context, or just individual contexts for each download?